sticht/metric.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2019 Yelp Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import threading
+import traceback
+from typing import Any
+from typing import Callable
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+import pytimeparse
+
+from sticht.slack import SlackDeploymentProcess
+
+
+# TODO: Rethink if MetricDemultiplexer is required
+
+
+class MetricWatcher:
+    def __init__(
+        self,
+        metric: Any,
+        callback: Callable[['MetricWatcher'], Any],
+        start_timestamp: float,
+        label: str,
+        max_duration: float,
+    ) -> None:
+        self.metric = metric
+        self.window: List[Tuple[float, float]] = []
+        self.max_duration = max_duration
+        self.failing: Optional[bool] = None
+        self.bad_after_mark: Optional[bool] = None
+        self.bad_before_mark: Optional[bool] = None
+        self.callback = callback
+        self.start_timestamp = start_timestamp
+        self.label = label
+
+    def process_datapoint(self, props, datapoint, timestamp) -> None:
+        self.window.append((timestamp, datapoint))
+        self.trim_window()
+
+        if timestamp > self.start_timestamp:
+            self.bad_after_mark = self.is_window_bad()
+        else:
+            self.bad_before_mark = self.is_window_bad()
+
+        old_failing = self.failing
+        self.failing = self.bad_after_mark and not self.bad_before_mark
+
+        if self.failing == (not old_failing):
+            self.callback(self)
+
+    def trim_window(self) -> None:
+        old_window = self.window
+        min_ts = max(ts for ts, d in old_window) - self.window_duration()
+        new_window = [(ts, d) for (ts, d) in old_window if ts > min_ts]
+        self.window = new_window
+
+    def window_duration(self) -> float:
+        """Many of our SLOs are defined with durations of 1 hour or more; this is great if you're trying to avoid being
```

Comment:
minor nit on myself: but do we want to keep this? This is still currently true about our SLOs.